lib/history/sqlite.go

```diff
@@ -0,0 +1,379 @@
+	events := diffCluster(t.clock, t.lastStatus, status)
+	if len(events) == 0 {
+		return nil
+	}
+
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	tx, err := t.database.BeginTx(ctx, nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	defer func() {
+		// The rollback will be ignored if the tx has already been committed.
+		if err == nil {
+			return
+		}
+		if err := tx.Rollback(); err != nil {
+			log.WithError(err).Error("Failed to rollback sql transaction.")
+		}
+	}()
+
+	if err = t.insertEvents(ctx, tx, events); err != nil {
+		return trace.Wrap(err, "failed to insert events")
+	}
+
+	if t.size+len(events) > t.capacity {
+		if err = t.evictEvents(ctx, tx); err != nil {
+			return trace.Wrap(err, "failed to evict old events")
+		}
+	}
+
+	if err = tx.Commit(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	t.size = t.size + len(events)
+	if t.size > t.capacity {
+		t.size = t.capacity
+	}
+
+	t.lastStatus = status
+	return nil
+}
+
+// GetEvents returns a filtered list of events based on the provided params.
+func (t *SQLiteTimeline) GetEvents(ctx context.Context, params map[string]string) (events []*pb.TimelineEvent, err error) {
+	query, args := prepareQuery(params)
+	rows, err := t.database.QueryxContext(ctx, query, args...)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	defer func() {
+		if err := rows.Close(); err != nil {
+			log.WithError(err).Error("Failed to close sql rows.")
+		}
+	}()
+
+	for rows.Next() {
+		var row sqlEvent
+		if err = rows.StructScan(&row); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		event, err := row.toProto()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		events = append(events, event)
+	}
+
+	if err := rows.Err(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return events, nil
+}
+
+// insertEvents inserts the provided events into the timeline.
+func (t *SQLiteTimeline) insertEvents(ctx context.Context, tx *sql.Tx, events []*pb.TimelineEvent) error {
+	for _, event := range events {
+		row, err := newSQLEvent(event)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		args := []interface{}{
+			row.Timestamp,
+			row.EventType,
+			row.Node.String,
+			row.Probe.String,
+			row.Old.String,
+			row.New.String,
+		}
+
+		if _, err := tx.ExecContext(ctx, insertIntoEvents, args...); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// evictEvents deletes oldest events if the timeline is larger than its max
+// capacity.
+func (t *SQLiteTimeline) evictEvents(ctx context.Context, tx *sql.Tx) error {
+	if _, err := tx.ExecContext(ctx, deleteOldFromEvents, t.capacity); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// prepareQuery prepares a query string and a list of arguments constructed from
+// the provided params.
+func prepareQuery(params map[string]string) (query string, args []interface{}) {
+	var fields = []string{"type", "node", "probe", "old", "new"}
+	var sb strings.Builder
+	index := 0
+
+	sb.WriteString("SELECT * FROM EVENTS ")
```

Comment:
Nit: I'm curious if you tried to do this using text/template approach? Nothing wrong with using string builder, just wondering if that'd be a bit more cleaner code-wise.